File: apps/go/internal/chapter15/validation.go
package chapter15

import (
	"fmt"
	"slices"
	"strconv"
	"strings"
)

// Validated は検証の結果。Go には sealed interface が無いので、
// 理由の一覧が空かどうかで「正しい」「不正」を表す構造体にする。
type Validated[T any] struct {
	Value  T
	Errors []string
}

// Valid は理由が 1 つも無いかどうか。
func (v Validated[T]) Valid() bool {
	return len(v.Errors) == 0
}

// validate は理由（空文字列は問題なし）を集め、1 つも無ければ値を作る。
func validate[T any](reasons []string, value func() (T, error)) (Validated[T], error) {
	errs := make([]string, 0, len(reasons))

	for _, reason := range reasons {
		if reason != "" {
			errs = append(errs, reason)
		}
	}

	if len(errs) > 0 {
		return Validated[T]{Errors: errs}, nil
	}

	built, err := value()
	if err != nil {
		return Validated[T]{}, err
	}

	return Validated[T]{Value: built}, nil
}

// required は値が JSON に無ければ理由を返す。
func required(field string, present bool) string {
	if present {
		return ""
	}

	return field + " は必須です"
}

// notNegative は値が負であれば理由を返す。値が無ければ何も言わない（required の担当）。
func notNegative[N float64 | int](field string, value *N) string {
	if value == nil || *value >= 0 {
		return ""
	}

	return field + " は 0 以上にしてください"
}

// oneOf は値が選択肢に無ければ理由を返す。値が無ければ何も言わない。
func oneOf[E comparable](field string, value *E, allowed []E) string {
	if value == nil || slices.Contains(allowed, *value) {
		return ""
	}

	choices := make([]string, len(allowed))
	for i, choice := range allowed {
		choices[i] = fmt.Sprintf("%v", choice)
	}

	slices.Sort(choices)

	return field + " は " + strings.Join(choices, "、") + " のどれかにしてください"
}

// MovieRequest は興行収入の予測の要求。
// Go の数値は 0 が既定値なので、JSON に値が無かったことを表すためにポインタで受ける。
type MovieRequest struct {
	SNS1     *float64 `json:"sns1"`
	SNS2     *float64 `json:"sns2"`
	Actor    *float64 `json:"actor"`
	Original *int     `json:"original"`
}

// originalValues は原作の有無に使える値。
var originalValues = []int{0, 1}

// Validate は検証して、正しければ映画の特徴量にする。
func (r MovieRequest) Validate() (Validated[Movie], error) {
	return validate(
		[]string{
			required("sns1", r.SNS1 != nil),
			required("sns2", r.SNS2 != nil),
			required("actor", r.Actor != nil),
			required("original", r.Original != nil),
			notNegative("sns1", r.SNS1),
			notNegative("sns2", r.SNS2),
			notNegative("actor", r.Actor),
			oneOf("original", r.Original, originalValues),
		},
		func() (Movie, error) {
			return Movie{SNS1: *r.SNS1, SNS2: *r.SNS2, Actor: *r.Actor, Original: *r.Original}, nil
		},
	)
}

// PassengerRequest は生存の予測の要求。年齢と乗船港は省略できる。
type PassengerRequest struct {
	Pclass   *int     `json:"pclass"`
	Sex      *string  `json:"sex"`
	Age      *float64 `json:"age"`
	SibSp    *int     `json:"sib_sp"`
	Parch    *int     `json:"parch"`
	Fare     *float64 `json:"fare"`
	Embarked *string  `json:"embarked"`
}

var (
	// passengerClasses は客室の等級に使える値。
	passengerClasses = []int{1, 2, 3}
	// sexes は性別に使える値。
	sexes = []string{"female", "male"}
	// ports は乗船港に使える値。
	ports = []string{"C", "Q", "S"}
)

// Validate は検証して、正しければ乗客の特徴量にする。省略された列は空文字列（欠損値）にする。
func (r PassengerRequest) Validate() (Validated[Passenger], error) {
	return validate(
		[]string{
			required("pclass", r.Pclass != nil),
			required("sex", r.Sex != nil),
			required("sib_sp", r.SibSp != nil),
			required("parch", r.Parch != nil),
			required("fare", r.Fare != nil),
			oneOf("pclass", r.Pclass, passengerClasses),
			oneOf("sex", r.Sex, sexes),
			notNegative("age", r.Age),
			notNegative("sib_sp", r.SibSp),
			notNegative("parch", r.Parch),
			notNegative("fare", r.Fare),
			oneOf("embarked", r.Embarked, ports),
		},
		func() (Passenger, error) {
			return Passenger{
				Pclass:   strconv.Itoa(*r.Pclass),
				Sex:      *r.Sex,
				Age:      optionalNumber(r.Age),
				SibSp:    strconv.Itoa(*r.SibSp),
				Parch:    strconv.Itoa(*r.Parch),
				Fare:     formatNumber(*r.Fare),
				Embarked: optionalText(r.Embarked),
			}, nil
		},
	)
}

// optionalNumber は省略できる数値をセルの文字列にする。省略されていれば空文字列（欠損値）。
func optionalNumber(value *float64) string {
	if value == nil {
		return ""
	}

	return formatNumber(*value)
}

// formatNumber は数値をセルの文字列にする。
func formatNumber(value float64) string {
	return strconv.FormatFloat(value, 'f', -1, 64)
}

// optionalText は省略できる文字列をセルにする。
func optionalText(value *string) string {
	if value == nil {
		return ""
	}

	return *value
}

File: apps/clojure/src/getting_started_ml/chapter15/validation.clj
(ns getting-started-ml.chapter15.validation
  "第 15 章のプレゼンテーション層。要求の JSON を読み、検証してドメインの値にする。

   Cheshire は JSON の値をそのまま Clojure のマップ・数値・文字列にするので、
   Java 版のように「読み込みと同時に型が確かめられる」ことはない。型の確認は自分で書く。"
  (:require [cheshire.core :as json]
            [clojure.string :as str])
  (:import [com.fasterxml.jackson.core JsonProcessingException]))

(def movie-types
  "興行収入の予測の要求の列と型。原作の有無は整数で受け取る（1.5 を弾くため）。"
  {"sns1" :number "sns2" :number "actor" :number "original" :integer})

(def passenger-types
  "生存の予測の要求の列と型。年齢と乗船港は省略できる。"
  {"pclass" :integer "sex" :string "age" :number "sib_sp" :integer
   "parch" :integer "fare" :number "embarked" :string})

(defn- typed?
  "値が型に合うかどうか。null と、表に無い列（型が nil）は問わない。"
  [type value]
  (case type
    nil true
    :number (or (nil? value) (number? value))
    :integer (or (nil? value) (integer? value))
    :string (or (nil? value) (string? value))))

(defn read-json
  "本文を JSON のオブジェクトとして読み、列の型を確かめる。読めないか型が合わなければ nil を返す。"
  [body types]
  (try
    (let [request (json/parse-string body)]
      (when (and (map? request)
                 (every? (fn [[field value]] (typed? (get types field) value)) request))
        request))
    (catch JsonProcessingException _ nil)))

;; 検証の規則。問題が無ければ nil を、あれば理由を返す

(defn required
  "必須の列が空なら理由を返す。"
  [field value]
  (when (nil? value) (str field " は必須です")))

(defn not-negative
  "負の数なら理由を返す。"
  [field value]
  (when (and (some? value) (neg? value)) (str field " は 0 以上にしてください")))

(defn one-of
  "選択肢の外の値なら理由を返す。選択肢は並べ替えて表示する。"
  [field value allowed]
  (when (and (some? value) (not (contains? (set allowed) value)))
    (str field " は " (str/join "、" (map str (sort allowed))) " のどれかにしてください")))

(defn validate
  "理由（nil は問題なし）を集め、1 つも無ければ f で値を作る。
   Java 版の sealed interface Validated と違い、値と理由を両方持つマップで表す。"
  [reasons f]
  (let [errors (vec (remove nil? reasons))]
    {:value (when (empty? errors) (f)) :errors errors}))

(defn valid?
  "検証の結果が正しいかどうか。"
  [validated]
  (empty? (:errors validated)))

(def ^:private original-values [0 1])
(def ^:private passenger-classes [1 2 3])
(def ^:private sexes ["female" "male"])
(def ^:private ports ["C" "Q" "S"])

(defn movie
  "検証して、正しければ映画の特徴量にする。"
  [request]
  (let [{:strs [sns1 sns2 actor original]} request]
    (validate [(required "sns1" sns1)
               (required "sns2" sns2)
               (required "actor" actor)
               (required "original" original)
               (not-negative "sns1" sns1)
               (not-negative "sns2" sns2)
               (not-negative "actor" actor)
               (one-of "original" original original-values)]
              #(hash-map :sns1 (double sns1) :sns2 (double sns2)
                         :actor (double actor) :original (long original)))))

(defn passenger
  "検証して、正しければ乗客の特徴量にする。年齢と乗船港は省略できる。"
  [request]
  (let [{:strs [pclass sex age parch fare embarked]} request
        sib-sp (get request "sib_sp")]
    (validate [(required "pclass" pclass)
               (required "sex" sex)
               (required "sib_sp" sib-sp)
               (required "parch" parch)
               (required "fare" fare)
               (one-of "pclass" pclass passenger-classes)
               (one-of "sex" sex sexes)
               (not-negative "age" age)
               (not-negative "sib_sp" sib-sp)
               (not-negative "parch" parch)
               (not-negative "fare" fare)
               (one-of "embarked" embarked ports)]
              #(hash-map :pclass (long pclass) :sex sex
                         :age (when age (double age))
                         :sib-sp (long sib-sp) :parch (long parch) :fare (double fare)
                         :embarked embarked))))
